src/common/meta/src/cache/table/composite_table_route.rs

```diff
@@ -0,0 +1,274 @@
+            CompositeTableRoute::Physical(route) => route,
+            CompositeTableRoute::Logical(_, route) => route,
+        }
+    }
+
+    /// Returns [LogicalTableRouteValue] reference if it's [CompositeTableRoute::Logical]; Otherwise returns [None].
+    pub fn as_logical_table_route_ref(&self) -> Option<&Arc<LogicalTableRouteValue>> {
+        match self {
+            CompositeTableRoute::Physical(_) => None,
+            CompositeTableRoute::Logical(route, _) => Some(route),
+        }
+    }
+}
+
+/// [CompositeTableRouteCache] caches the [TableId] to [CompositeTableRoute] mapping.
+pub type CompositeTableRouteCache = CacheContainer<TableId, Arc<CompositeTableRoute>, CacheIdent>;
+
+pub type CompositeTableRouteCacheRef = Arc<CompositeTableRouteCache>;
+
+/// Constructs a [CompositeTableRouteCache].
+pub fn new_composite_table_route_cache(
+    name: String,
+    cache: Cache<TableId, Arc<CompositeTableRoute>>,
+    table_route_cache: TableRouteCacheRef,
+) -> CompositeTableRouteCache {
+    let init = init_factory(table_route_cache);
+
+    CacheContainer::new(name, cache, Box::new(invalidator), init, Box::new(filter))
+}
+
+fn init_factory(
+    table_route_cache: TableRouteCacheRef,
+) -> Initializer<TableId, Arc<CompositeTableRoute>> {
+    Arc::new(move |table_id| {
+        let table_route_cache = table_route_cache.clone();
+        Box::pin(async move {
+            let table_route_value = table_route_cache
+                .get(*table_id)
+                .await?
+                .context(error::ValueNotExistSnafu)?;
+            match table_route_value.as_ref() {
+                TableRoute::Physical(physical_table_route) => Ok(Some(Arc::new(
+                    CompositeTableRoute::Physical(physical_table_route.clone()),
+                ))),
+                TableRoute::Logical(logical_table_route) => {
+                    let physical_table_id = logical_table_route.physical_table_id();
+                    let physical_table_route = table_route_cache
+                        .get(physical_table_id)
+                        .await?
+                        .context(error::ValueNotExistSnafu)?;
+
+                    let physical_table_route = physical_table_route
+                        .as_physical_table_route_ref()
+                        .with_context(|| error::UnexpectedSnafu {
+                            err_msg: format!(
+                                "Expected the physical table route, but got logical table route, table: {table_id}"
+                            ),
+                        })?;
+
+                    Ok(Some(Arc::new(CompositeTableRoute::Logical(
+                        logical_table_route.clone(),
+                        physical_table_route.clone(),
+                    ))))
+                }
+            }
+        })
+    })
+}
+
+fn invalidator<'a>(
+    cache: &'a Cache<TableId, Arc<CompositeTableRoute>>,
+    ident: &'a CacheIdent,
+) -> BoxFuture<'a, Result<()>> {
```

Comment:
why not write an `async fn`

Reply:
The `CacheContainer` requires the signature `Box<dyn for<'a> Fn(&'a Cache<K, V>, &'a CacheToken) -> BoxFuture<'a, Result<()>> + Send + Sync>;`